File: 22_Java_Integration/SbtProject/src/main/scala/04_scala_option_in_java/ScalaOptionInJava.scala
package r04_scala_option_in_java

import scala.jdk.OptionConverters.*


// pp. 656-657
// create java.util.Optional[String] values
object Scala:
    // convert a Some to Optional
    val scalaStringSome = Option("foo").toJava

    // convert a None to Optional
    val scalaStringNone = Option.empty[String].toJava

    // p. 657
    val intOptional1 = Option(1).toJava            // Optional[Int], or 
                                                   // Optional[Object]
    val optionalInt = Option(1).toJavaPrimitive    // OptionalInt
    val optionalInt2 = optionalInt.toJavaGeneric   // Optional[Int]

    // p. 658
    val optionString = Option("foo")

File: 22_Java_Integration/SbtProject/src/main/java/04_scala_option_in_java/ScalaOptionInJava.java
package r04_scala_option_in_java;

// for the 1st example
import java.util.Optional;

// for the 2nd example
import java.util.OptionalInt;

// for the 3rd example
import scala.jdk.javaapi.OptionConverters;

public class ScalaOptionInJava {
    public static void main(String[] args) {

        // p. 656
        Optional<String> stringSome = Scala.scalaStringSome();  // Optional[foo]
        Optional<String> stringNone = Scala.scalaStringNone();  // Optional.empty

        // p. 657
        Optional intOptional1   = Scala.intOptional1();   // Optional[1]
        OptionalInt optionalInt = Scala.optionalInt();    // OptionalInt[1]
        Optional optionalInt2   = Scala.optionalInt2();   // Optional[1]

        // p. 658
        Optional<String> stringOptional = OptionConverters.toJava(Scala.optionString());
        
    }
}



    